smithy-model/src/main/java/software/amazon/smithy/model/traits/HttpChecksumProperty.java

```diff
@@ -0,0 +1,236 @@
+        this.algorithm = Objects.requireNonNull(builder.algorithm);
+        this.location = Objects.requireNonNull(builder.location);
+        this.name = Objects.requireNonNull(builder.name);
+    }
+
+    public static HttpChecksumProperty.Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Create a {@code HttpChecksumProperty} from {@link Node}.
+     *
+     * @param node {@code Node} to create the {@code HttpChecksumProperty}
+     * @return Returns the created {@code HttpChecksumProperty}.
+     * @throws ExpectationNotMetException if the given {@code node} is invalid.
+     */
+    public static HttpChecksumProperty fromNode(Node node) {
+        ObjectNode value = node.expectObjectNode().warnIfAdditionalProperties(KEYS);
+        Builder builder = builder();
+
+        value.getStringMember(ALGORITHM).map(StringNode::getValue).ifPresent(builder::algorithm);
+        value.getStringMember(LOCATION).map(StringNode::getValue).ifPresent(builder::location);
+        value.getStringMember(NAME).map(StringNode::getValue).ifPresent(builder::name);
+        return builder.build();
+    }
+
+    /**
+     * @return Returns the supported checksum algorithm.
+     */
+    public String getAlgorithm() {
+        return algorithm;
+    }
+
+    /**
+     * @return Returns a location supported for the checksum algorithm.
+     */
+    public Location getLocation() {
+        return location;
+    }
+
+    /**
+     * @return Returns header or trailer name to be used with checksum algorithm.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Take this object and create a builder that contains all of the
+     * current property values of this object.
+     *
+     * @return Returns a builder for type HttpChecksumProperties.
+     */
+    @Override
+    public SmithyBuilder<HttpChecksumProperty> toBuilder() {
+        return builder()
+                .algorithm(getAlgorithm())
+                .location(getLocation())
+                .name(getName());
+    }
+
+    /**
+     * Converts a value to a {@link Node}.
+     *
+     * @return Returns the creates Node.
+     */
+    @Override
+    public Node toNode() {
+        return Node.objectNodeBuilder()
+                .withMember(ALGORITHM, getAlgorithm())
+                .withMember(LOCATION, getLocation().toNode())
+                .withMember(NAME, getName())
+                .build();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, location, algorithm);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (!(o instanceof HttpChecksumProperty)) {
+            return false;
+        }
+
+        HttpChecksumProperty that = (HttpChecksumProperty) o;
+        return algorithm.equals(that.algorithm)
+                && location.equals(that.location);
```

Comment:
No, this is intended.  We consider checksum property equal when they are modeled for the same checksum algorithm and location. 

This is because modeling same checksum algorithm and location is not valid.  

I use equals to raise an error when we encounter multiple values with same checksum algorithm and location.

Reply:
I raise an error if duplicate entries are found while adding in request list [1] or response list [2] 

[1] https://github.com/awslabs/smithy/pull/843/files#diff-c1d6f5ffdbf17d22712737361d41edb190675537fe90920e3844f1d7afeae8bfR181

[2] https://github.com/awslabs/smithy/pull/843/files#diff-c1d6f5ffdbf17d22712737361d41edb190675537fe90920e3844f1d7afeae8bfR203

Reply:
`.equals` should be checking for true equality. You can always add a separate method that detects if there's a conflict, we do [this with `UriPattern`](https://github.com/awslabs/smithy/blob/076547fea82e04491c490b1e0675a6340900ae6f/smithy-model/src/main/java/software/amazon/smithy/model/pattern/UriPattern.java#L128).